multithreaded/Dfit_2D_main.cpp

```diff
@@ -74,211 +77,297 @@ int main(int argc, char *argv[])
 	std::cout << nentries << " number of entries read"  <<std::endl;
 	for (int i=0; i<nentries; ++i){
 		ch.GetEntry(i);
-		if (mu_PT < muPTmin || mu_P < muPmin || mu_eta < eta_min || mu_eta > eta_max)
-			continue;
-		double B_MMcorr = B_M + 2.0 * missPT;
-		if (B_MMcorr < minBMcorr || B_MMcorr > maxBMcorr)
+		if (!Config::isMC){
+			if (mu_PT < Config::muPTmin || mu_P < Config::muPmin || mu_eta < Config::eta_min || mu_eta > Config::eta_max)
+				continue;
+			if (int(charge) != Config::sign)
+				continue;
+			B_MMcorr = B_M + 2.0 * missPT;
+		}
+		if (B_MMcorr < Config::minBMcorr || B_MMcorr > Config::maxBMcorr)
 			continue;
-		if (D_M < minDM || D_M > maxDM)
+			
+		if (D_M < Config::minDM || D_M > Config::maxDM)
 			continue;
-		if (int(charge) == sign)
-		{
-			vect_2D.push_back(std::make_pair(D_M, B_MMcorr));
-		}
+		vect_2D.push_back(std::make_pair(D_M, B_MMcorr));
+		hist2D.Fill(D_M, B_MMcorr);
+		
+		if (Config::isMC)
+			frac_indeces[frac_index]++;
+		
 	}
+	std::cout<< "++++++++++++++++++++++++++++++++++++++++++++\n";
+	std::cout << hist2D.Integral() << "  " << vect_2D.size() << std::endl;
 
 
 	// Initial fit parameter values taken from 1D fits to MC and Side Bands
 
 		
 
 	// Define Minuit fit variables for M_D
-
-	const auto& contrName = Config::contrName;
-	const auto& varname_md = Config::varname_md;
-
-	double starting_point[(nvar_md+nvar_mb)*ncontr+ncontr-1];
-	for (int i = 0; i < ncontr; i++)
-	{
-		for (int ivar = 0; ivar < nvar_md; ivar++)
+	const int n_all = (Config::nvar_md+Config::nvar_mb) * Config::ncontr + Config::ncontr;
+	bool previous_fit = false;
+	double starting_point[n_all];
+	std::string minName = "Minuit2";
+	std::string algoName = "";
+	int itry = 1;
+	bool goodfit = false;
+while ((itry<=Config::ntries || !goodfit)&&itry<=100){
+	bool start_scratch = true;
+	int loop_count = 0;
+	for (auto& int_choose_fit: Config::int_choose_fits){
+		ROOT::Math::Minimizer *min =
+			ROOT::Math::Factory::CreateMinimizer(minName, algoName);
+
+		// Set tolerance , etc...
+		min->SetMaxFunctionCalls(Config::functionCalls); // for Minuit/Minuit2
+		min->SetTolerance(Config::tolerance[loop_count]);
+		loop_count++;
+		min->SetPrintLevel(Config::printLevel);
+		// min->SetStrategy(2);
+
+		for (int i = 0; i < Config::ncontr; i++)
 		{
-			min->SetVariable(i * nvar_md + ivar, (TString::Format("%s_%s", contrName[i].c_str(), varname_md[ivar].c_str())).Data(), MC_MD[i][ivar], dMC_MD[i][ivar] + 1.0e-11);
-			starting_point[i * nvar_md + ivar] = MC_MD[i][ivar];
+			for (int ivar = 0; ivar < Config::nvar_md; ivar++)
+			{
+				min->SetVariable(i * Config::nvar_md + ivar, (TString::Format("%s_%s", Config::contrName[i].c_str(), Config::varname_md[ivar].c_str())).Data(), Config::MC_MD[i][ivar], Config::dMC_MD[i][ivar] + 1.0e-11);
+				if (((int_choose_fit == dictionaryChooseFit.at("BM")|| int_choose_fit == dictionaryChooseFit.at("frac")) && i!=-1)){
+		       			min->FixVariable(i * Config::nvar_md + ivar);
+				}else{
+					if (ivar>4) 
```

Comment:
To oraz poniżej `ivar==3`.

Myślę że czystsze byłoby wczytanie z configu które parametry są fixed

Zwłaszcza że już to robisz w linii 268